src/plugins/discover/public/application/main/components/layout/use_discover_histogram.ts

```diff
@@ -6,277 +6,258 @@
   );
 
+  useEffect(() => {
+    unifiedHistogram?.setRequestParams({
+      dataView,
+      query,
+      filters,
+      timeRange,
+      searchSessionId,
+      requestAdapter: inspectorAdapters.requests,
+    });
+  }, [
+    dataView,
+    filters,
+    inspectorAdapters.requests,
+    query,
+    searchSessionId,
+    timeRange,
+    unifiedHistogram,
+  ]);
+
   /**
-   * Total hits
+   * Override Unified Histgoram total hits with Discover partial results
    */
 
-  const [localHitsContext, setLocalHitsContext] = useState<UnifiedHistogramHitsContext>();
-
-  const onTotalHitsChange = useCallback(
-    (status: UnifiedHistogramFetchStatus, result?: number | Error) => {
-      if (result instanceof Error) {
-        // Display the error and set totalHits$ to an error state
-        sendErrorTo(data, savedSearchData$.totalHits$)(result);
-        return;
-      }
-
-      const { fetchStatus, recordRawType } = savedSearchData$.totalHits$.getValue();
-
-      // If we have a partial result already, we don't want to update the total hits back to loading
-      if (fetchStatus === FetchStatus.PARTIAL && status === UnifiedHistogramFetchStatus.loading) {
-        return;
-      }
-
-      // Set a local copy of the hits context to pass to unified histogram
-      setLocalHitsContext({ status, total: result });
+  const firstLoadComplete = useRef(false);
 
-      // Sync the totalHits$ observable with the unified histogram state
-      savedSearchData$.totalHits$.next({
-        fetchStatus: status.toString() as FetchStatus,
-        result,
-        recordRawType,
-      });
-
-      // Check the hits count to set a partial or no results state
-      if (status === UnifiedHistogramFetchStatus.complete && typeof result === 'number') {
-        checkHitCount(savedSearchData$.main$, result);
-      }
-    },
-    [data, savedSearchData$.main$, savedSearchData$.totalHits$]
-  );
-
-  // We only rely on the totalHits$ observable if we don't have a local hits context yet,
-  // since we only want to show the partial results on the first load, or there will be
-  // a flickering effect as the loading spinner is quickly shown and hidden again on fetches
-  const { fetchStatus: hitsFetchStatus, result: hitsTotal } = useDataState(
+  const { fetchStatus: totalHitsStatus, result: totalHitsResult } = useDataState(
     savedSearchData$.totalHits$
   );
 
-  const hits = useMemo(
-    () =>
-      isPlainRecord
-        ? undefined
-        : localHitsContext ?? {
-            status: hitsFetchStatus.toString() as UnifiedHistogramFetchStatus,
-            total: hitsTotal,
-          },
-    [hitsFetchStatus, hitsTotal, isPlainRecord, localHitsContext]
-  );
+  useEffect(() => {
+    // We only want to show the partial results on the first load,
+    // or there will be a flickering effect as the loading spinner
+    // is quickly shown and hidden again on fetches
+    if (!firstLoadComplete.current) {
+      unifiedHistogram?.setTotalHits({
```

Comment:
The frequency of `unifiedHistogram?` seems to be an unpleasant consequence of the api initialization pattern. This code will fail silently if `unifiedHistogram` is `undefined` - this shouldn't happen, but if it does it'll be difficult to debug. It would be extremely nice if the optional chaining could be removed. If `unifiedHistogram` is `undefined`, I want it to fail FAST.

Reply:
Yeah, I believe this will be a consequence of any pattern we use to pass API objects up from components, whether that be through refs or callback props -- mainly due to the fact you need to declare a state variable up front to hold the API object which will be by definition `undefined` until the component hands it to you.

In these cases though, the API being `undefined` isn't necessarily a failure condition. It's ok for these effects to run without the API being defined yet, and having no effect in these cases is the intended behaviour. I originally had guard clauses to explicitly terminate these effects early if the API was `undefined`, but I personally found these more cumbersome than using `?.`, so I got rid of them.

With that said, I'd agree that not being able to understand when the API is `undefined` because `initialize` _was not yet called_ vs cases where `initialize` _will never be called_ (or failed) is absolutely a flaw with this pattern. It's a point that was raised to me in a discussions about the pros and cons of refs vs callback props for passing the API to the consumer, and is an area where the callback prop approach is probably superior (unless the consumer-provided callback itself fails, which leads to a similar situation).

>would it be possible to get the unifiedHistogram as a promise? There may be places where you'd still choose to use it as you do here but you'd at least have the option of using await

I think I understand the thinking here, since the promise would never resolve to `undefined`, but it would probably pose similar issues and potentially new ones. If I `await` the promise but never call `initialize`, presumably the promise will never resolve, which I think would be similarly confusing. It would also force us to grapple with the use of async code in a React component, which is a whole other can of worms.